pytorch_lightning/plugins/training_type/full_sharded.py

```diff
@@ -0,0 +1,150 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List, Optional
+
+import torch
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.plugins.training_type.ddp import DDPPlugin
+from pytorch_lightning.utilities import _FAIRSCALE_AVAILABLE, _FAIRSCALE_FULL_SHARDED_AVAILABLE
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if _FAIRSCALE_AVAILABLE:
+    from fairscale.nn.data_parallel import FullyShardedDataParallel
+
+    from pytorch_lightning.overrides.fairscale import (
+        LightningFullShardedDataParallel,
+        unwrap_lightning_module_full_sharded,
+    )
+
+
+class FullShardedPlugin(DDPPlugin):
+
+    def __init__(
+        self,
+        cpu_offload: bool = True,
+        flatten_parameters: bool = False,
+        reshard_after_forward: bool = True,
+        move_grads_to_cpu: Optional[bool] = None,
+        fp32_reduce_scatter: Optional[bool] = None,
+        compute_dtype: Optional[torch.dtype] = None,
+        bucket_cap_mb: int = 25,
+        parallel_devices: Optional[List[torch.device]] = None,
+        num_nodes: int = 1,
+        cluster_environment: ClusterEnvironment = None,
+        sync_batchnorm: Optional[bool] = False
+    ):
+        """
+
+               Provides capabilities to run training using the Full Sharded capabilities provided by FairScale.
+
+               Full Sharded Training shards the entire model across all available GPUs, allowing you to scale model
+               size, whilst using efficient communication to reduce overhead. In practice, this means we can remain
+               at parity with PyTorch DDP, whilst scaling our model sizes dramatically. The technique is similar
+               to ZeRO-Stage 3 but have been modified/adjusted for PyTorch.
+
+               `For more information: https://fairscale.readthedocs.io/en/latest/api/nn/fsdp.html`.
+
+               .. warning:: ``FullShardedPlugin`` is in beta and subject to change.
+
+               Defaults have been set to enable CPU Offload, but options have been exposed and may require configuration
+               based on your level of memory/speed efficiency.
+               We suggest having a look at this PR for more information.
+               `https://github.com/facebookresearch/fairscale/pull/413`
+
+
+               Many of the helpful doc strings below came from the original FairScale documentation:
+               `https://fairscale.readthedocs.io/en/latest/api/nn/fsdp.html`
+
+               Arguments:
+
+                   cpu_offload: Offload FP32 params to CPU. Only useable in precision=16 mode (default: False).
```

Comment:
Reminder to add a check for precision=16 if cpu_offload

Reply:
for cpu_offload, the model should start out residing on the CPU, not GPU AFAIK. I don't know if you have that already.

Reply:
huge thanks for checking @min-xu-ai :) we do not move to to CUDA if using CPU offload, just running into other fp16 scaler issues regarding cpu offload